### PR TITLE
Use the median-of-3 when gathering timings

### PR DESCRIPTION
**Pull request by ldionne**
With this PR, the timings are done 3 times for each `n` in the range, and the median of the 3 timings is taken. This is an experiment to check whether the timings appear to be more reliable this way, as discussed in #69. Note that running the benchmarks now takes 3 times as long, which could be a deal breaker.

### CODE REVIEW

**Comment by brunocodutra** [comment]
Looks good, did it improve the reproducibility of benchmark results?

**Comment by ldionne** [comment]
> Looks good, did it improve the reproducibility of benchmark results?

Haven't checked yet. I'm trying to think of a way to compare reproducibility, since just running the benchmarks with and without the median trick won't tell us much. I think this would work: we could generate the benchmarks multiple times without the median trick, and then multiple times with the median trick. Then, we would look at the average dispersion of the curves with and without the trick. Does that make sense?

**Comment by brunocodutra** [comment]
> We could generate the benchmarks multiple times without the median trick, and then multiple times with the median trick. Then, we would look at the average dispersion of the curves with and without the trick. Does that make sense?

That is exactly what I mean.

Also, the trick could be made into a feature, i.e. `metabench_add_dataset` could take an additional parameter that defaults to `1` that sets the number of times each compilation should be repeated. We could even decide to enable the trick for MPL and Fusion only and perhaps even set a higher number of repetitions for them (specially for Fusion).

File: metabench.cmake [hunk 1]
```diff
@@ -247,6 +247,11 @@ file(WRITE ${METABENCH_RB_PATH}
 "require 'time'                                                                                         \n"
 "                                                                                                       \n"
 "module Metabench                                                                                       \n"
+"  def self.median(values)                                                                              \n"
+"    values = values.sort                                                                               \n"
+"    return values[values.length / 2]                                                                   \n"
```
Comment (brunocodutra) [inline metabench.cmake]: When `values.length` is even, the median should be `(values[values.length / 2 - 1] + values[values.length / 2])/2`

**Comment by ldionne** [comment]
Adding a customization point adds complexity, and I'm really not sure of the added value. Also, if we go this way, I don't see a reason not to allow deciding whether we pick the median or the mean of the n observations, which would add even more complexity. I think we must be careful not to over-engineer the data acquisition, since this is a very deep hole.

File: metabench.cmake (hunk 1 above)
Comment (ldionne) [inline metabench.cmake]: Fixed.

**Comment by brunocodutra** [comment]
> think we must be careful not to over-engineer the data acquisition, since this is a very deep hole.
View changes

I agree, but in this case it could be useful to reduce the time required for running the benchmarks and thus avoid blowing up Travis limits. Anyway, lets first assess whether it improves reproducibility.

**Comment by ldionne** [comment]
At first glance, it's a bit difficult to tell whether the median trick really gives an improvement. I run the benchmarks 11 times with and without the median trick, and here are the generated results.

The 11 runs with the median trick:  [median.zip](https://github.com/ldionne/metabench/files/198829/median.zip)
The 11 runs without the median trick: [no_median.zip](https://github.com/ldionne/metabench/files/198830/no_median.zip)

Now, I might have been using my computer during some of these benchmarks (nothing intensive, though), but then again the benchmarks are run on Travis and the cores can be shared, so we should be able to deal with that.

I think the general trend is that the median trick does improve the results, in the sense that we rarely get curves that are completely foobar.

**Comment by ldionne** [comment]
If we do provide this feature, it should be optional because that'll blow up the benchmark times, and that's not acceptable.

**Comment by ldionne** [comment]
Also note that the difference seems to be much greater for benchmarks of heterogeneous computations than for pure type computations.

**Comment by ldionne** [comment]
So the bottom line is that it's slightly better (it couldn't worsen things, since it's a median), but the benefit may not be so huge. Do we want to make the data acquisition for this benefit? Maybe, I'm not sure. I'll wait for your input.

**Comment by brunocodutra** [comment]
>  run the benchmarks 11 times with and without the median trick, and here are the generated results.

Now that's what I call professionalism.

> I'll wait for your input.

In fact I think the improvement is quite noticeable and I'm sure it would get increasingly better the larger the size of the median filter, so I think it is definitely worth it to provide the feature even if we might not be able to take full advantage of it for the benchmark suite due to Travis constraints. Remember that others will hopefully find Metabench useful for their needs and a median filter might come in handy.

In any case it should be quite easy to provide the feature from this point, so why not?

**Comment by ldionne** [comment]
I would say this fixes #69.

**Comment by brunocodutra** [comment]
Looks great.

**Pull request merged**